transport/session/src/types.rs

```diff
@@ -238,76 +249,107 @@ pub struct IncomingSession {
     pub target: SessionTarget,
 }
 
-// TODO: missing docs
+/// Represents the Session protocol socket over HOPR.
+///
+/// This is essentially a HOPR-specific wrapper for [`ReliableSocket`] and [`UnreliableSocket`]
+/// Session protocol sockets.
+#[pin_project::pin_project]
 pub struct Session {
     id: SessionId,
-    inner: Pin<Box<dyn AsyncReadWrite>>,
+    #[pin]
+    inner: Box<dyn AsyncReadWrite>,
     routing: DestinationRouting,
     capabilities: Capabilities,
-    on_close: Option<Box<dyn FnOnce(SessionId) + Send + Sync>>,
+    on_close: Option<Box<dyn FnOnce(SessionId, ClosureReason) + Send + Sync>>,
 }
 
 impl Session {
-    pub fn new(
+    /// Creates a new HOPR Session.
+    ///
+    /// It builds an [`futures::io::AsyncRead`] + [`futures::io::AsyncWrite`] transport
+    /// from the given `hopr` interface and passing it to the appropriate [`UnreliableSocket`] or [`ReliableSocket`]
+    /// based on the given `capabilities`.
+    #[tracing::instrument(skip(hopr, on_close), fields(session_id = %id))]
+    pub fn new<Tx, Rx, C>(
         id: SessionId,
         routing: DestinationRouting,
-        capabilities: Capabilities,
-        tx: Arc<dyn SendMsg + Send + Sync>,
-        rx: Pin<Box<dyn futures::Stream<Item = Box<[u8]>> + Send + Sync>>,
-        on_close: Option<Box<dyn FnOnce(SessionId) + Send + Sync>>,
-    ) -> Self {
-        let inner_session = InnerSession::new(id, routing.clone(), tx, rx);
-
-        // If we request any capability, we need to use Session protocol
-        if !capabilities.is_empty() {
-            // This is a very coarse assumption, that 3-hop takes at most 3 seconds.
-            // We can no longer base this timeout on the number of hops because
-            // it is not known for SURB-based routing.
-            let rto_base = Duration::from_secs(3);
-
-            let expiration_coefficient =
-                if !capabilities.is_disjoint(Capability::RetransmissionAck | Capability::RetransmissionNack) {
-                    4
-                } else {
-                    1
-                };
+        capabilities: C,
+        hopr: (Tx, Rx),
+        on_close: Option<Box<dyn FnOnce(SessionId, ClosureReason) + Send + Sync>>,
+    ) -> Result<Self, TransportSessionError>
+    where
+        Tx: futures::Sink<(DestinationRouting, ApplicationData)> + Send + Sync + Unpin + 'static,
+        Rx: futures::Stream<Item = Box<[u8]>> + Send + Sync + Unpin + 'static,
```

Comment:
We will also have to solve this ugly asymmetry of different types of data objects passed in and out.